src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -1,207 +1,375 @@
-
-                if (_receiver != null && _receiver.IsValueCreated)
-                {
-                    Receiver.CloseAsync().Wait();
-                    _receiver = null;
-                }
-
-                if (_clientEntity != null)
-                {
-                    _clientEntity.CloseAsync().Wait();
-                    _clientEntity = null;
-                }
-
-                _disposed = true;
-            }
-        }
-
+
+        private IMessageSession _messageSession;
+        private SessionMessageProcessor _sessionMessageProcessor;
+
+        private Lazy<ServiceBusScaleMonitor> _scaleMonitor;
+
+        public ServiceBusListener(string functionId, EntityType entityType, string entityPath, bool isSessionsEnabled, ITriggeredFunctionExecutor triggerExecutor,
+            ServiceBusOptions config, ServiceBusAccount serviceBusAccount, MessagingProvider messagingProvider, ILoggerFactory loggerFactory, bool singleDispatch)
+        {
+            _functionId = functionId;
+            _entityType = entityType;
+            _entityPath = entityPath;
+            _isSessionsEnabled = isSessionsEnabled;
+            _triggerExecutor = triggerExecutor;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _messagingProvider = messagingProvider;
+            _serviceBusAccount = serviceBusAccount;
+            _loggerFactory = loggerFactory;
+            _logger = loggerFactory.CreateLogger<ServiceBusListener>();
+            _receiver = CreateMessageReceiver();
+            _scaleMonitor = new Lazy<ServiceBusScaleMonitor>(() => new ServiceBusScaleMonitor(_functionId, _entityType, _entityPath, _serviceBusAccount.ConnectionString, _receiver, _loggerFactory));
+            _singleDispatch = singleDispatch;
+
+            if (_isSessionsEnabled)
+            {
+                _sessionMessageProcessor = _messagingProvider.CreateSessionMessageProcessor(_entityPath, _serviceBusAccount.ConnectionString);
+            }
+            else
+            {
+                _messageProcessor = _messagingProvider.CreateMessageProcessor(entityPath, _serviceBusAccount.ConnectionString);
+            }
+            _serviceBusOptions = config;
+        }
+
+        internal MessageReceiver Receiver => _receiver.Value;
+
+        internal IMessageSession MessageSession => _messageSession;
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+
+            if (_started)
+            {
+                throw new InvalidOperationException("The listener has already been started.");
+            }
+
+            if (_receiver == null)
+            {
+                _receiver = CreateMessageReceiver();
+            }
+
+            if (_isSessionsEnabled)
+            {
+                _clientEntity = _messagingProvider.CreateClientEntity(_entityPath, _serviceBusAccount.ConnectionString);
+
+                if (_singleDispatch)

```

Comment:
Looks like when singleDispatch is false you do the same thing here, regardless of whether sessions is enabled. So it'd be good to hoist this out to it's own, e.g.

```csharp
if (_singleDispatch)
{
    if (_sessionsEnabled)
    {
    }
    else
    {
    }
}
else
{
}
```